File: 1-working-with-list/p25_permutation.scala
import scala.util.Random;

object App {

  def at[T](i: Int, l: List[T]): Option[T] = {
    l match {
      case Nil => None
      case h :: t => if(i == 0) Some(h) else at(i-1, t)
    }
  }

  def removeAt[T](i: Int, lst: List[T]): List[T] = {
    def aux[T](acc: List[T], cur: Int, l: List[T]): List[T] = {
      l match {
        case Nil => acc
        case h :: t => if(cur == 0) acc ::: t else aux(h::acc, cur-1, t)
      }
    }
    aux(List(), i, lst)
  }

  def permutation[T](lst: List[T]): List[T] = {
    def aux[T](acc: List[T], l: List[T]): List[T] = {
      l match {
        case Nil => acc
        case _ =>
          val r = Random.nextInt(l.size)
          val e = at(r, l) match { case Some(x) => x; case None => throw new Exception() }
          val sl = removeAt(r, l)
          aux(e::acc, sl)
      }
    }
    aux(List(), lst)
  }

  def main(args: Array[String]) {
    val l = List("a", "b", "c", "d", "e" ,"f")
    println(l)
    println(permutation(l))
  }
}

File: 1-working-with-list/p25_permutation.ml
let rec at n l =
  match l with
  | [] -> None
  | hd :: tail when n > 0 -> if n = 1 then Some hd else at (n-1) tail
  | _ -> None ;;

let remove_at n l =
  let rec aux acc curr = function
  | [] -> List.rev acc
  | h :: t ->
    if curr = 0 then
      (List.rev acc) @ t
    else
      aux (h::acc) (curr-1) t
  in aux [] n l;;

let permutation lst =
  let rec aux acc = function
    | [] -> acc
    | l ->
      let r = Random.int (List.length l) in
      let e = match (at (r+1) l) with Some(x) -> x | _ -> assert false in
      let sl = remove_at r l in
      aux (e::acc) sl
  in aux [] lst;;

permutation ["a"; "b"; "c"; "d"; "e"; "f"];;
